### PR TITLE
[stdlib] Add ContiguousCollection

### PR DESCRIPTION
**Pull request by airspeedswift**
Implementation for apple/swift-evolution#990

Tests tk

### CODE REVIEW

File: stdlib/public/core/ContiguousCollection.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A collection that supports access to its underlying contiguous storage.
+@available(iOS 9999, OSX 9999, tvOS 9999, watchOS 9999, *)
```
Comment (benrimmington) [inline stdlib/public/core/ContiguousCollection.swift]:
```suggestion
@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
```

`OSX` => `macOS` (11 times) and in same order as rest of stdlib, for easier find and replace.

File: stdlib/public/core/ContiguousCollection.swift [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+public protocol ContiguousCollection: Collection
+where SubSequence: ContiguousCollection {
+  /// Calls a closure with a pointer to the array's contiguous storage.
```
Comment (benrimmington) [inline stdlib/public/core/ContiguousCollection.swift]:
```suggestion
  /// Calls a closure with a pointer to contiguous storage.
```

Remove: `the array's` (also in the proposal).

File: stdlib/public/core/ContiguousCollection.swift [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    return try withContiguousStorageIfAvailable(body)!
+  }
+}
+
+/// A collection that supports mutable access to its underlying contiguous
+/// storage.
+@available(iOS 9999, OSX 9999, tvOS 9999, watchOS 9999, *)
+public protocol MutableContiguousCollection: ContiguousCollection, MutableCollection
+where SubSequence: MutableContiguousCollection {
+  /// Calls the given closure with a pointer to the array's mutable contiguous
```
Comment (benrimmington) [inline stdlib/public/core/ContiguousCollection.swift]:
```suggestion
  /// Calls the given closure with a pointer to mutable contiguous
```

Remove: `the array's` (also in the proposal).

File: stdlib/public/core/ContiguousCollection.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    return try withContiguousStorageIfAvailable(body)!
+  }
+}
+
+/// A collection that supports mutable access to its underlying contiguous
+/// storage.
+@available(iOS 9999, OSX 9999, tvOS 9999, watchOS 9999, *)
+public protocol MutableContiguousCollection: ContiguousCollection, MutableCollection
```
Comment (benrimmington) [inline stdlib/public/core/ContiguousCollection.swift]: Wrap to 80 characters?

File: stdlib/public/core/ContiguousCollection.swift [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+extension UnsafeBufferPointer: ContiguousCollection { }
+
+@available(iOS 9999, OSX 9999, tvOS 9999, watchOS 9999, *)
+extension UnsafeMutableBufferPointer: MutableContiguousCollection { }
+
+@available(iOS 9999, OSX 9999, tvOS 9999, watchOS 9999, *)
+extension Slice: ContiguousCollection where Base: ContiguousCollection { }
+
+@available(iOS 9999, OSX 9999, tvOS 9999, watchOS 9999, *)
+extension Slice: MutableContiguousCollection where Base: MutableContiguousCollection { }
```
Comment (benrimmington) [inline stdlib/public/core/ContiguousCollection.swift]: Wrap to 80 characters?

**Comment by tkremenek** [comment]
SE-0256 has been rejected during formal Swift Evolution review.  Closing this PR.

**Pull request closed without merging**